BUG: Tolerant intersection

cpelley opened the pull request:
This PR stops intersection from wrapping grids for the following situation:

The grid is being split in two at either edge due to the point being a value of (base + period).

This is not at all a neat solution but it yield the desired result.  Any alternate solutions are welcome.

Further context in #1662

cpelley commented:
Not sure why the travis build failed.

pelson commented:
Thanks @cpelley - looks like you've got a solution which fits all of the existing tests and fits your requirements, nicely done!
I'm very conscious that this has added a *lot* of complexity to the determination of subsets, which essentially means that this will be the last time that anybody can fully grasp the subset computation comfortably in their head. For that reason, I think this needs a whole lot more commenting, documentation of the overall *aim* of subsets, and I would suggest splitting some of the indexing lines into sensibly named (aka self documenting) variable definitions.
Because of the complexity, I currently cannot be certain that you have tested all of the cases that you have added, so I will defer reviewing the tests until the logic is a little easier to review.

:+1: in principle on finding a solution to the problem, and I'm happy to help moving it forwards with this PR.

cpelley commented:
Thanks for your comment @pelson.  I'll make changes based on your comment ASAP.

cpelley commented:
Done, cheers @pelson, ready for review.

cpelley commented:
ping :)

cpelley commented:
ping x2
Is there anything stopping this PR from getting a look into?

rhattersley commented on lib/iris/cube.py:
```diff
@@ -2278,8 +2282,68 @@ def _intersect_modulus(self, coord, minimum, maximum, min_inclusive,
+                                             inside_indices[-1] + 1)]
+                else:
+                    subsets = [
+                        slice(index_of_second_chunk, None),
```
This line changed during your development but doesn't seem to have been any corresponding change in the tests - perhaps it hasn't been covered by the tests? Or perhaps the tests didn't always pass?

rhattersley replied:
NB. This code path *is* tested so the tests probably broke at some point.

rhattersley commented on lib/iris/cube.py:
```diff
@@ -2278,8 +2282,68 @@ def _intersect_modulus(self, coord, minimum, maximum, min_inclusive,
+                                        coord.points[0] + modulus):
+                            points[-1] = coord.points[-1]
+                            subsets = [slice(inside_indices[0],
+                                       inside_indices[-1] + 1)]
```
This code path isn't tested.

cpelley replied:
Thanks @rhattersley

cpelley replied:
I managed to write a test for this that was not actually testing it - I have corrected this now.
Out of interest how did you spot this?  Did you but break points here like I just did?

rhattersley commented on lib/iris/cube.py:
```diff
@@ -2278,8 +2282,68 @@ def _intersect_modulus(self, coord, minimum, maximum, min_inclusive,
+                                        coord.points[-1] + modulus):
+                            points[0] = coord.points[0]
+                            subsets = [slice(inside_indices[0],
+                                             inside_indices[-1] + 1)]
```
This code path isn't tested.

cpelley replied:
Thank you @rhattersley

rhattersley commented:
This is still a complex, hard-to-follow change so a lot of my confidence in this change is going to come from the testing. (Hence the comments about untested code paths.)

It'd be great if `Cube._intersect_modulus` could be re-factored to break it down into "bite-size" functions... but if you don't have time for that then I'm OK with merging as it stands.

rhattersley commented:
> It'd be great if Cube._intersect_modulus could be re-factored

For example:
```
$ flake8 --max-complexity=10 lib/iris/cube.py
...
lib/iris/cube.py:2204:1: C901 'Cube._intersect_modulus' is too complex (17)
...
```

cpelley commented:
Thank you @rhattersley for your time on reviewing this.  I totally agree about the complexity, hence my description in the PR.  I'll certainly ensure complete test coverage as you have quite rightly suggested.  With regards to refactoring, I'll have a look to see if there are quick refactors possible without too much controversy.  I'll open a new ticket to capture this requirement if I don't think the refactor is complete (it certainly needs it as I found it difficult enough to follow the logic before I introduced my changes, let alone now).

`flake8` - Nice, I didn't know about this tool.  I'll read up about what heuristics it uses to determine complexity.  Sounds VERY useful.

Cheers @rhattersley, I'll get back to you ASAP

cpelley commented:
I have added a test and modified an existing one so now both paths are tested.

Now looking at the possibility of a refactor, will let you know if I get anywhere ASAP

cpelley commented:
I have a made I think a pretty good refactor (makes a lot more sense now).

Over to you, cheers @rhattersley

cpelley commented:
Apologies, forgot to push back up.  Ready now.

cpelley commented:
You happy with the refactor @rhattersley ?

I think intersection should eventually be pulled out into its own module but I have approach the refactor from the perspective of increasing code clarity, simplification and removing duplication.  I think this approach is most important right now.

rhattersley commented on lib/iris/cube.py:
```diff
@@ -2197,6 +2201,90 @@ def create_coords(src_coords, add_coord):
+            Derive the subset for dimension coordinates.
+
+            Ensure that we do not wrap if blocks are at the very edge.  That
+            is, if the very edge if wrapped and corresponds to base + period,
```
Typo: "edge **if** wrapped" -> "edge **is** wrapped"

rhattersley commented:
Thanks for the refactor. It's a definite improvement.

If you fix the typo and squash the commits I'm :+1: to merge.

rhattersley commented:
Oops. Just realised it could do with a [what's new entry](https://github.com/SciTools/iris/tree/master/docs/iris/src/whatsnew/contributions_1.9). Feel free to leave it as a separate commit - whatever's easier.

cpelley commented:
Done, cheers @rhattersley
